### PR TITLE
Possible fixes #8 compilation failed regex too large

### LINKED ISSUE #8: preg_match(): Compilation failed: regular expression is too large at offset 156184
This happens while using `Zend\Validator\Hostname->isValid('tapi4457@hsoqvf.biz')` because hostname/Biz.php contains enormous regular expression

Trace looks like this:

preg_match('/^[\\x{002d}0-9a-z\\x{3447}\\x{3473}\\x{359..(156088 chars skipped)..9FA0}\\x{9FA2}\\x{9FA4}\\x{9FA5}]{1,20}$/iu','tapi4457@hsoqvf')
  /vendor/zendframework/zend-validator/src/Hostname.php:1877

PHP Version 7.1.7

---

Originally posted by @ac-dc at https://github.com/zendframework/zend-validator/issues/219

### PR DESCRIPTION
**Pull request by samsonasik**
Signed-off-by: Abdul Malik Ikhsan <samsonasik@gmail.com>

|    Q          |   A
|-------------- | ------
| Bugfix        | yes

### Description

This is a possible quick fix for #8 . I am not sure if it is correct, but hostname with `@` seems invalid.

### CODE REVIEW

**Comment by samsonasik** [comment]
@ac-dc @Thaix could you verify if this patch works for you ?

**Comment by Thaix** [comment]
Yes, it works on my machine!
Before:

```
$ php -a
Interactive mode enabled

php > include 'vendor/autoload.php';
php > var_dump((new Laminas\Validator\Hostname())->isValid('tapi4457@hsoqvf.biz'));
PHP Warning:  preg_match(): Compilation failed: regular expression is too large at offset 156184 in /home/apidev/current/vendor/laminas/laminas-validator/src/Hostname.php on line 2104
bool(false)
```

After
```
php -a
Interactive mode enabled

php > include 'vendor/autoload.php';
php > var_dump((new Laminas\Validator\Hostname())->isValid('tapi4457@hsoqvf.biz'));
bool(false)
```

File: src/Hostname.php [hunk 1]
```diff
@@ -1977,6 +1977,11 @@ public function isValid($value)
             return true;
         }
 
+        if (strpos($value, '@') !== false) {
```
Comment (weierophinney) [inline src/Hostname.php:1983]: Please add a comment above this one indicating:

- The issue it resolves
- What the root cause is that it's resolving (`@` in a hostname is always invalid, so there's no need to use a regex to detect that scenario).

**Review by weierophinney — COMMENTED** [review]
This technically solves the reported issue. I'm a bit worried that we'll still see issues crop up around the "regexp compilation failed" due to regexp size, though. I wonder if it might be good to try and break it into a number of smaller `strpos()` checks for common, known-bad characters, which would allow us to reduce what the regexp looks for?

File: src/Hostname.php (hunk 1 above)
Comment (samsonasik) [reply src/Hostname.php:1983]: I've added the comment.

**Comment by samsonasik** [comment]
@weierophinney do we have list of known-bad characters for hostname?

**Comment by weierophinney** [comment]
> do we have list of known-bad characters for hostname?

I think part of the complexity of the regex is a list of characters to exclude, IIRC. :smile:

**Comment by samsonasik** [comment]
How about merge for now?

**Comment by samsonasik** [comment]
@weierophinney I think we can merge as is for now, as concrete solution seems not imaginable right now.

**Pull request merged**